### PR TITLE
Quarkus Super Heroes Workshop S2I tests

### PR DESCRIPTION
**Pull request by mjurc**
* Adding tests verifying that Quarkus Super Heroes Workshop resources work when deployed through RHBQ supported S2I deployment.
* Moving `todo-demo-app` to `external-applications` folder.

### CODE REVIEW

File: external-applications/quarkus-workshop-super-heroes/pom.xml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    <artifactId>quarkus-workshop-super-heroes</artifactId>
+    <packaging>jar</packaging>
+
+    <name>Quarkus OpenShift TS: Workshop Super Heroes</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy</artifactId>
+        </dependency>
```
Comment (Ladicek) [inline external-applications/quarkus-workshop-super-heroes/pom.xml:23]: Do you need this dep?

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@OpenShiftTest
+@ManualDeployment(appName = "quarkus-workshop-hero", httpRoot = "/", knownEndpoint = "/")
+@AdditionalResources("classpath:openjdk-11-rhel7.yaml")
+@AdditionalResources("classpath:heroes-database.yaml")
+@AdditionalResources("classpath:hero.yaml")
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
```
Comment (Ladicek) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:30]: You're not using the `@Order` annotation :-)

Comment (Ladicek) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:30]: Actually you are, but not on all tests. Weird, but if it works for you, then sure :-)

File: external-applications/quarkus-workshop-super-heroes/pom.xml (hunk 1 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/pom.xml:23]: I am just using the constants from WS API jar, I should probably just use that actually

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        StringBuilder sb = new StringBuilder();
+        sb.append("{");
+        if (id != null && !id.isEmpty()) { sb.append("\"id\":" + id + ","); }
+        sb.append("\"name\":\"" + name + "\"");
+        sb.append(", \"otherName\":\"" + otherName + "\"");
+        sb.append(", \"level\":" + level);
+        sb.append(", \"picture\":\"" + picture + "\"");
+        sb.append(", \"powers\":\"" + powers + "\"");
+        sb.append("}");
+        return sb.toString();
```
Comment (Ladicek) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:172]: Use Jackson or JSON-P?

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@OpenShiftTest
+@ManualDeployment(appName = "quarkus-workshop-villain", httpRoot = "/", knownEndpoint = "/")
+@AdditionalResources("classpath:openjdk-11-rhel7.yaml")
+@AdditionalResources("classpath:villains-database.yaml")
+@AdditionalResources("classpath:villain.yaml")
+public class VillainsOpenShiftIT {
```
Comment (Ladicek) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java:30]: Here, you use `@Order`, but you're missing `@TestMethodOrder(MethodOrderer.OrderAnnotation.class)` :-)

File: external-applications/quarkus-workshop-super-heroes/pom.xml (hunk 1 above)
Comment (Ladicek) [reply external-applications/quarkus-workshop-super-heroes/pom.xml:23]: Ah OK, no, that makes sense.

**Comment by Ladicek** [comment]
We should figure out how to share some of those YAML files, but that's not for this PR. Looks fine to me, but I only skimmed through it.

File: external-applications/quarkus-workshop-super-heroes/pom.xml (hunk 1 above)
Comment (Ladicek) [reply external-applications/quarkus-workshop-super-heroes/pom.xml:23]: Perhaps just add it to `<scope>test</scope>`.

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java (hunk 4 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java:33]: Whoops. Fixed!

File: external-applications/quarkus-workshop-super-heroes/pom.xml (hunk 1 above)
Comment (rsvoboda) [reply external-applications/quarkus-workshop-super-heroes/pom.xml:24]: https://github.com/quarkusio/quarkus/blob/master/bom/runtime/pom.xml#L1612
You can add:
```xml
            <dependency>
                <groupId>org.jboss.spec.javax.ws.rs</groupId>
                <artifactId>jboss-jaxrs-api_2.1_spec</artifactId>
                <scope>test</scope>
            </dependency>
```

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java (hunk 2 above)
Comment (rsvoboda) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:30]: I think it's ok, test without annotation are independent

those 3 annotated need to go in create - update - delete order

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+import org.junit.jupiter.api.TestMethodOrder;
+
+import java.net.URL;
+
+import static io.restassured.RestAssured.given;
+import static io.restassured.RestAssured.when;
+import static javax.ws.rs.core.HttpHeaders.ACCEPT;
+import static javax.ws.rs.core.HttpHeaders.CONTENT_TYPE;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.Response.Status.*;
```
Comment (rsvoboda) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:20]: I don't like `*` imports, just few codes are used

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                .statusCode(OK.getStatusCode());
+    }
+
+    @Test
+    public void testMetrics() {
+        given()
+                .header(ACCEPT, APPLICATION_JSON)
+                .when().get(url + "/metrics/application")
+                .then()
+                .statusCode(OK.getStatusCode());
```
Comment (rsvoboda) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:91]: can you do check on certain metrics being exposed?

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        given()
+                .body(rb)
+                .header(CONTENT_TYPE, APPLICATION_JSON)
+                .header(ACCEPT, APPLICATION_JSON)
+                .when()
+                .put("/api/heroes")
+                .then()
+                .statusCode(OK.getStatusCode())
+                .header(CONTENT_TYPE, APPLICATION_JSON)
+                .body("name", Is.is(UPDATED_NAME))
```
Comment (rsvoboda) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:144]: `Is.is` => `is` aka static import

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java (hunk 3 above)
Comment (rsvoboda) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:172]: an example can be found here - https://javaee.github.io/jsonp/getting-started.html

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+import org.junit.jupiter.api.TestMethodOrder;
+
+import javax.ws.rs.core.HttpHeaders;
+
+import java.net.URL;
+
+import static io.restassured.RestAssured.given;
+import static io.restassured.RestAssured.when;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.Response.Status.*;
```
Comment (rsvoboda) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java:20]: please remove `*` imports

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+	private static final String UPDATED_PICTURE = "hackerman.png";
+	private static final String DEFAULT_POWERS = "Partakes in this test";
+	private static final String UPDATED_POWERS = "Partakes in update test";
+	private static final int DEFAULT_LEVEL = 42;
+	private static final int UPDATED_LEVEL = 43;
+
+	@Test
+	public void testHello() {
+		when()
+				.get(url + "/api/villains/hello")
```
Comment (rsvoboda) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java:54]: the offset looks too much, one or additional \tab comparing to `HeroesOpenShiftIT`

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+		given()
+				.pathParam("id", villainId)
+				.when().delete("/api/villains/{id}")
+				.then()
+				.statusCode(NO_CONTENT.getStatusCode());
+	}
+
+	private static String getVillainJson(String id, String name, String otherName, String level, String picture,
+									  String powers) {
+		StringBuilder sb = new StringBuilder();
```
Comment (rsvoboda) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java:165]: JSON-P pls

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java (hunk 3 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java]: Actually, seems like [body](https://github.com/rest-assured/rest-assured/blob/master/rest-assured/src/main/java/io/restassured/specification/RequestSpecification.java#L151) method in RestAssured takes care of serialization, so I decided to use that.

Comment (Ladicek) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java]: That's even better, for some reason I forgot about it :-)

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java (hunk 9 above)
Comment (Ladicek) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java:54]: Yea, tab indentation is evil and should be banned for eternity. Spaces only!

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java (hunk 5 above)
Comment (Ladicek) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java]: Agree.

File: external-applications/quarkus-workshop-super-heroes/pom.xml (hunk 1 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/pom.xml]: Fixed.

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java (hunk 5 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java]: Fixed.

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java (hunk 6 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:92]: Check done at:

- https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/28/files#diff-e70bcb24b82b4e87bcce957bf5b39d16R171
- https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/28/files#diff-424ef8a617a8061b706b69289202f2fbR172

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java (hunk 7 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java]: Fixed.

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java (hunk 8 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java]: Fixed.

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java (hunk 9 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java]: This is hopefully fixed too now :)

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java (hunk 10 above)
Comment (mjurc) [reply external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/VillainsOpenShiftIT.java]: Fixed.

**Comment by mjurc** [comment]
@rsvoboda @Ladicek I think all of the problems are now resolved, would you kindly re-review? :slightly_smiling_face:

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+            .header(ACCEPT, APPLICATION_JSON)
+            .when().get(url + "/metrics/application")
+            .then()
+            .statusCode(OK.getStatusCode())
+            .body("'io.quarkus.workshop.superheroes.hero.HeroResource.countCreateHero'", is(1))
+            .body("'io.quarkus.workshop.superheroes.hero.HeroResource.countUpdateHero'", is(1))
+            .body("'io.quarkus.workshop.superheroes.hero.HeroResource.countDeleteHero'", is(1));
+    }
+
+    class Hero {
```
Comment (Ladicek) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:184]: This should be a `static class`, it doens't need to hold a reference to the test class.

File: external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+@OpenShiftTest
+@ManualDeployment(appName = "quarkus-workshop-hero", httpRoot = "/", knownEndpoint = "/")
+@AdditionalResources("classpath:openjdk-11-rhel7.yaml")
+@AdditionalResources("classpath:heroes-database.yaml")
+@AdditionalResources("classpath:hero.yaml")
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+public class HeroesOpenShiftIT {
+
+    @TestResource
+    private URL url;
```
Comment (Ladicek) [inline external-applications/quarkus-workshop-super-heroes/src/test/java/io/quarkus/ts/openshift/heroes/workshop/HeroesOpenShiftIT.java:35]: This should IMHO be below the `static` fields. The constants should be first, then the `heroId` (where we probably should add a comment that it's a workaround for passing data between test methods; alternatively, we could make `heroId` an instance field and make the test `@TestInstance(Lifecycle.PER_CLASS)`), and then this instance field.

**Comment by Ladicek** [comment]
Just added a few cosmetic comments, otherwise good work!

**Pull request merged**